### PR TITLE
For #7364 - Adds sessionId to DownloadState and populates it in DowloadFeature

### PR DESCRIPTION
**Pull request by codrut-topliceanu**
---

For https://github.com/mozilla-mobile/android-components/issues/7364

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks.
- [x] **Tests**: This PR doesn't include tests as it is only a minor change.
- [x] **Changelog**: This PR does not need [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md).
- [x] **Accessibility**: The code in this PR does not include any user facing features.

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by codrut-topliceanu** [comment]
Some unit tests are failing. :( Looking into them now.

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt [hunk 1]
```diff
@@ -93,9 +93,8 @@ class DownloadsFeature(
             flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(tabId) }
                 .ifChanged { it.content.download }
                 .collect { state ->
-                    val download = state.content.download
-                    if (download != null) {
-                        processDownload(state, download)
+                    state.content.download?.let { downloadState ->
```
Comment (Amejia481) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: I think will be better if we populate this property when the download request is issued https://github.com/mozilla-mobile/android-components/blob/9e12caf058910ed0a2b1d5fa78c8ffdb01b6ff1c/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/ContentStateReducer.kt#L65

This way we are going to have a consistent state in `BrowserStore`. what do you think?

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: @Amejia481 Thanks for taking a look at this. That's a really good solution that hadn't occurred to me! I just tried it now and it works great. Thank you.

Comment (csadilek) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: @Amejia481 @codrut-topliceanu Thanks and sorry for the late response. 

When creating a `DownloadState` we always know the `sessionId` (it should be called `tabId` in browser-state, btw). So, we can simply set the `tabId` here when creating the `download`: https://github.com/mozilla-mobile/android-components/blob/master/components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt#L185

It doesn't need to be a side-effect / unexpected change of reducing.

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: I already tried that and it didn't work, I'm afraid. It only worked for download that had skip confirmation == true.

Comment (csadilek) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: @codrut-topliceanu Hm can you share some more details? It effectively shouldn't matter when the sessionId is added, so why not add it directly when the download is created? It seems like a potential source of error to do this "later".

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: I'll look some more into it now, but I remember that it just wouldn't populate the sessionId in `EngineObserver` when `onExternalResource` was called (except when it was a larger download and it needed a confirmation from the user).

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: @csadilek Correction: `EngineObserver.onExternalResource` doesn't get called at all if the download is small enough that it doesn't need to show the confirmation dialog. I'm not entirely sure why,  or how to proceed.

Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: I think the source event is on [here](https://github.com/mozilla-mobile/android-components/blob/28c9be258747ff964d0a92808a57ae31f791cb31/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt#L667), if you debug that line you should able to see why is not getting called.

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: Yep that's the source. Can't go further up the tree from that as it leads into Gecko.

Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: Do you mean that when you try to download a file in some cases this [line](https://github.com/mozilla-mobile/android-components/blob/28c9be258747ff964d0a92808a57ae31f791cb31/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt#L667) does get executed?  If that it's the case then this is a GeckoView issue and it's a separate issue from this one.

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: Yep that's the case.

**Comment by codrut-topliceanu** [comment]
bors r=csadilek

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt (hunk 1 above)
Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: Would you mind opening a [ bug in Bugzilla](https://bugzilla.mozilla.org/enter_bug.cgi?product=GeckoView) describing the the issue and the step to reproduce it?

Comment (codrut-topliceanu) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: Ok, I'll give it a try. I've never opened one in Bugzilla before.

Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt:96]: If you need any help just let me know, happy to help!

**Pull request merged**